palace/linalg/errorestimator.cpp

```diff
@@ -0,0 +1,395 @@
+        iodata.solver.linear.mg_max_levels, iodata.solver.linear.mg_legacy_transfer,
+        iodata.solver.pa_order_threshold, mesh, smooth_flux_fecs)),
+    smooth_projector(smooth_flux_fes, iodata.solver.linear.tol, 200, 0,
+                     iodata.solver.pa_order_threshold),
+    coarse_flux_fec(iodata.solver.order, mesh.back()->Dimension(),
+                    mfem::BasisType::GaussLobatto),
+    coarse_flux_fes(mesh.back().get(), &coarse_flux_fec, mesh.back()->Dimension()),
+    scalar_mass_matrices(fes.GetNE()), smooth_to_coarse_embed(fes.GetNE())
+{
+  mfem::MassIntegrator mass_integrator;
+  for (int e = 0; e < fes.GetNE(); e++)
+  {
+    // Loop over each element, and save an elemental mass matrix.
+    // Will exploit the fact that vector L2 mass matrix components are independent.
+    const auto &coarse_fe = *coarse_flux_fes.GetFE(e);
+    auto &T = *coarse_flux_fes.GetElementTransformation(e);
+    mass_integrator.AssembleElementMatrix(coarse_fe, T, scalar_mass_matrices[e]);
+
+    const auto &smooth_fe = *smooth_flux_fes.GetFinestFESpace().GetFE(e);
+    coarse_fe.Project(smooth_fe, T, smooth_to_coarse_embed[e]);
+  }
+}
+
+IndicatorsAndNormalization CurlFluxErrorEstimator::operator()(const ComplexVector &v,
+                                                              bool normalize) const
+{
+  mfem::ParComplexGridFunction field(&fes);
+  field.real().SetFromTrueDofs(v.Real());
+  field.imag().SetFromTrueDofs(v.Imag());
+  const int nelem = smooth_flux_fes.GetFinestFESpace().GetNE();
+
+  // Coefficients for computing the discontinuous flux., i.e. (W, μ⁻¹∇ × V).
+  CurlFluxCoefficient real_coef(field.real(), mat_op), imag_coef(field.imag(), mat_op);
+  auto rhs_from_coef = [](mfem::ParFiniteElementSpace &flux_fes, auto &coef)
+  {
+    Vector RHS(flux_fes.GetTrueVSize());
+
+    mfem::LinearForm rhs(&flux_fes);
+    rhs.AddDomainIntegrator(new VectorFEDomainLFIntegrator(coef));
+    rhs.UseFastAssembly(false);
+    rhs.Assemble();
+    flux_fes.GetProlongationMatrix()->MultTranspose(rhs, RHS);
+
+    return RHS;
+  };
+
+  const auto smooth_flux_rhs =
+      ComplexVector(rhs_from_coef(smooth_flux_fes.GetFinestFESpace(), real_coef),
+                    rhs_from_coef(smooth_flux_fes.GetFinestFESpace(), imag_coef));
+
+  // Given the RHS vector of non-smooth flux, construct a flux projector and perform mass
+  // matrix inversion in the appropriate space, giving f = M⁻¹ f̂.
+  auto build_flux = [](const FluxProjector &proj, const ComplexVector &flux_coef)
+  {
+    // Use a copy construction to match appropriate size.
+    ComplexVector flux(flux_coef);
+    proj.Mult(flux_coef, flux);
+    return flux;
+  };
+  auto smooth_flux = build_flux(smooth_projector, smooth_flux_rhs);
+
+  // Given a complex solution represented with a PetscParVector, build a ComplexGridFunction
+  // for evaluation.
+  auto build_func = [](const ComplexVector &f, mfem::ParFiniteElementSpace &fes)
+  {
+    mfem::ParComplexGridFunction flux(&fes);
```

Comment:
Again, this is another vector allocation that can be allocated once for all solves.